geoapps_utils/numerical.py

```diff
@@ -10,9 +10,94 @@
 from uuid import UUID
 
 import numpy as np
+from geoh5py import Workspace
+from geoh5py.objects import Curve
+from geoh5py.shared import Entity
+from scipy.interpolate import interp1d
 from scipy.spatial import Delaunay
 
 
+def densify_curve(curve: Curve, increment: float) -> np.ndarray:
+    """
+    Refine a curve by adding points along the curve at a given increment.
+
+    :param curve: Curve object to be refined.
+    :param increment: Distance between points along the curve.
+
+    :return: Array of shape (n, 3) of x, y, z locations.
+    """
+    locations = []
+    for part in curve.unique_parts:
+        logic = curve.parts == part
+        if logic is not None and curve.cells is not None and curve.vertices is not None:
+            cells = curve.cells[np.all(logic[curve.cells], axis=1)]
+            vert_ind = np.r_[cells[:, 0], cells[-1, 1]]
+            locs = curve.vertices[vert_ind, :]
+            locations.append(resample_locations(locs, increment))
+
+    return np.vstack(locations)
+
+
+def get_locations(workspace: Workspace, entity: UUID | Entity):
+    """
+    Returns entity's centroids or vertices.
+
+    If no location data is found on the provided entity, the method will
+    attempt to call itself on its parent.
+
+    :param workspace: Geoh5py Workspace entity.
+    :param entity: Object or uuid of entity containing centroid or
+        vertex location data.
+
+    :return: Array shape(*, 3) of x, y, z location data
+
+    """
+    locations = None
+
+    if isinstance(entity, UUID):
+        entity_obj = workspace.get_entity(entity)[0]
+    elif isinstance(entity, Entity):
+        entity_obj = entity
+    if entity_obj is None:
+        return None
+
+    if hasattr(entity_obj, "centroids"):
+        locations = entity_obj.centroids
+    elif hasattr(entity_obj, "vertices"):
+        locations = entity_obj.vertices
+    elif (
+        getattr(entity_obj, "parent", None) is not None
+        and entity_obj.parent is not None
```

Comment:
**question:** what the difference between testing though `getattr` and accessing the member directly? Seem redundant